crates/data_components/src/s3_vectors/query_provider.rs

```diff
@@ -146,38 +156,137 @@ impl TableProvider for S3VectorsQueryTable {
             .collect())
     }
 
+    #[allow(clippy::too_many_lines)]
     async fn scan(
         &self,
-        _state: &dyn Session,
+        state: &dyn Session,
         projection: Option<&Vec<usize>>,
         filters: &[Expr],
         limit: Option<usize>,
     ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
-        #[allow(clippy::cast_possible_wrap)]
-        let limit: i64 = match limit {
-            Some(l) if (l as i64) > S3_VECTOR_MAX_TOPK => {
-                tracing::warn!(
-                    "S3VectorsQueryTable: limit {l} exceeds maximum of {S3_VECTOR_MAX_TOPK}, truncating."
-                );
-                S3_VECTOR_MAX_TOPK
-            }
-            None => S3_VECTOR_MAX_TOPK,
-            Some(l) => l as i64,
-        };
-
         let query_vector = self
             .compute_vector
             .compute_vector(self.query.as_str())
             .await
             .map_err(DataFusionError::External)?;
 
-        Ok(Arc::new(S3VectorsQueryExec::new(
-            self,
-            projection,
-            limit,
-            query_vector,
-            filters.to_vec(),
-        )) as Arc<dyn ExecutionPlan>)
+        if self.partition_by.is_empty() {
+            #[allow(clippy::cast_possible_wrap)]
+            let limit: i64 = match limit {
+                Some(l) if (l as i64) > S3_VECTOR_MAX_TOPK => {
+                    tracing::warn!(
+                        "S3VectorsQueryTable: limit {l} exceeds maximum of {S3_VECTOR_MAX_TOPK}, truncating."
+                    );
+                    S3_VECTOR_MAX_TOPK
+                }
+                None => S3_VECTOR_MAX_TOPK,
+                Some(l) => l as i64,
+            };
+            let plan = Arc::new(S3VectorsQueryExec::new(
+                self,
+                projection,
+                limit,
+                query_vector,
+                filters.to_vec(),
+            ));
+
+            return Ok(plan as Arc<dyn ExecutionPlan>);
+        }
+
+        let (_, Some(bucket_name), Some(index_name)) = self.table.idx.index_identifier_variables()
+        else {
+            return exec_err!("No bucket name or index name for bucket query");
+        };
+
+        let list_indexes_output = self
+            .table
+            .client
+            .list_indexes(
```

Comment:
We have to discover the partitioned indexes every scan here because QueryTableProvider is created anew each invocation of `vector_search`.